File: skills/JuliusBrussee--caveman/packages/middleware/python/caveman_middleware/mcp.py
"""Native MCP result views and local host recovery registration.

The caller keeps its Client/ClientSession, transports and original results.
This adapter adds no MCP server or protocol implementation. All compression and
scoped recovery use the shared Engine-backed Caveman runtime.
"""
from __future__ import annotations

from collections.abc import Awaitable, Callable, Mapping, Sequence
from dataclasses import dataclass
from importlib.metadata import version
import json
import uuid
from typing import Any

try:
    from mcp.types import CallToolResult, TextContent, Tool
except ModuleNotFoundError as error:
    raise ImportError("Install caveman-middleware[mcp] for the native MCP adapter") from error

from caveman_cloud.middleware import Adapter, AsyncMiddlewareRuntime, Candidate, MiddlewareError, Scope, sha256
from ._native import owner
from ._versions import matches_framework


@dataclass(frozen=True)
class MCPToolBinding:
    """The native MCP definition and the callable registered in the host loop."""
    tool: Tool
    execute: Callable[..., Awaitable[CallToolResult]]


def bind_mcp_tool(client, tool: Tool) -> MCPToolBinding:
    """Use an existing native client; it keeps auth, IDs, validation and options."""
    async def execute(arguments=None, **native_options):
        return await client.call_tool(tool.name, arguments, **native_options)
    return MCPToolBinding(tool, execute)


class CavemanMCPHost:
    def __init__(self, *, runtime: AsyncMiddlewareRuntime, scope: Scope,
                 server_id: str, protocol_version: str):
        if not isinstance(runtime, AsyncMiddlewareRuntime):
            raise TypeError("Native MCP clients require AsyncMiddlewareRuntime")
        self._version_supported = matches_framework(("mcp", "2.2", "3"))
        if not self._version_supported and runtime.mode != "off":
            runtime.decline("unsupported_version")
        if not server_id or not protocol_version:
            raise ValueError("Provide the host's server identity and negotiated protocol version")
        self.runtime, self.scope, self.server_id = runtime, scope, server_id
        self.adapter = Adapter("mcp", "0.1.0", "2.2.0", "mcp-native-" + protocol_version + "-v1")
        binding = runtime.recovery(scope)
        self._binding = binding
        tool = Tool(name=binding.name, description=binding.description, input_schema=dict(binding.input_schema))

        async def execute(arguments=None, **_native_options):
            # This is a host-local executor, not an outbound MCP tools/call.
            page = await binding.execute(arguments or {})
            return CallToolResult(content=[TextContent(type="text", text=json.dumps(page, ensure_ascii=False, separators=(",", ":")))])
        self.recovery = MCPToolBinding(tool, execute)
        self._recovery_executor = execute
        self._recovery_definition = tool.model_dump_json(by_alias=True, exclude_none=True)

    def register(self, tools: Sequence[MCPToolBinding]) -> list[MCPToolBinding]:
        """Append only our executable tool. A name collision leaves tools intact."""
        if not self._version_supported or self.runtime.mode != "compress" or any(item.tool.name == self.recovery.tool.name for item in tools):
            return list(tools)
        return [*tools, self.recovery]

    async def project_result(self, result: CallToolResult, *, tool: Tool, call_id: str,
                             context_manifest: Sequence[Mapping[str, str]],
                             registered_tools: Sequence[MCPToolBinding] = (),
                             sequence: int | None = None) -> CallToolResult:
        """Build the model view immediately before dispatch. Persist result itself.

        context_manifest is the host's append-only original-context manifest.
        It must be reused across resumes; deliberate edits need a new cache epoch.
        This boundary sees MCP values, not the provider's final serialized prompt.
        """
        def skipped(reason):
            self.runtime.report(None, reason=reason, adapter=self.adapter.id)
            return result

        if owner.get() is not None:
            return result
        if not self._version_supported or self.runtime.mode == "off":
            return skipped("unsupported_version")
        if type(result) is not CallToolResult or type(tool) is not Tool:
            return skipped("unsupported_shape")
        if result.is_error or result.result_type != "complete" or "structured_content" in result.model_fields_set or tool.output_schema is not None or tool.name.startswith("caveman_"):
            return skipped("protected_result")
        candidates, indices = [], {}
        for i, part in enumerate(result.content):
            if type(part) is not TextContent or part.annotations and part.annotations.audience is not None and "assistant" not in part.annotations.audience:
                continue
            key = "mcp-" + sha256(json.dumps([self.server_id, tool.name, call_id, i], separators=(",", ":")))
            indices[key] = i
            candidates.append(Candidate(id=key, source_id=key, content=part.text))
        if not candidates:
            return skipped("no_candidate")

        def registered():
            matching = [item for item in registered_tools if item.tool.name == self._binding.name]
            return (self.runtime.owns_binding(self._binding, self.scope) and len(matching) == 1 and matching[0] is self.recovery and
                    self.recovery.execute is self._recovery_executor and
                    self.recovery.tool.model_dump_json(by_alias=True, exclude_none=True) == self._recovery_definition)

        bound = registered()
        outcome = await self.runtime.optimize(scope=self.scope, adapter=self.adapter, candidates=candidates,
                    manifest=context_manifest, sequence=sequence, binding=self._binding if bound else None,
                    recovery_overhead_text=self.recovery.tool.model_dump_json(by_alias=True, exclude_none=True),
                    logical_call_id=str(uuid.uuid4()), attempt_id=str(uuid.uuid4()))
        if not outcome.replacements:
            self.runtime.report(outcome)
            return result
        if bound and not registered():
            return skipped("recovery_unavailable")
        if not all(item["segment_id"] in indices for item in outcome.replacements):
            return skipped("invalid_plan")
        content = list(result.content)
        for replacement in outcome.replacements:
            index = indices[replacement["segment_id"]]
            content[index] = content[index].model_copy(update={"text": replacement["text"]})
        view = result.model_copy(update={"content": content})
        self.runtime.report(outcome)
        return view

File: skills/JuliusBrussee--caveman/packages/middleware/typescript/src/mcp.ts
import type { Client } from '@modelcontextprotocol/sdk/client/index.js';
import type { CallToolResult, Tool } from '@modelcontextprotocol/sdk/types.js';
import { MiddlewareRuntime, sha256, type ManifestItem, type Scope } from '@caveman-ai/sdk/middleware';
import { currentOwner, plain } from './common.js';
import { adapterCompatible, frameworkVersion } from './compatibility.js';

/** A native MCP definition and the callable actually registered in the host. */
export interface MCPToolBinding {
  tool: Tool;
  execute: (arguments_: Record<string, unknown>, options?: Parameters<Client['callTool']>[2]) => Promise<CallToolResult>;
}

/** Preserve the existing client's transport, IDs, auth and result validation. */
export function bindMCPTool(client: Client, tool: Tool): MCPToolBinding {
  return { tool, execute: (arguments_, options) => client.callTool({ name: tool.name, arguments: arguments_ }, undefined, options) as Promise<CallToolResult> };
}

export interface MCPHostOptions { runtime: MiddlewareRuntime; scope: Scope; serverId: string; protocolVersion: string }

/** Native host views only. No MCP framing, server or inference loop is added. */
export class CavemanMCPHost {
  readonly recovery: MCPToolBinding;
  private readonly binding;
  private readonly adapter;
  private readonly recoveryExecutor;
  private readonly recoveryDefinition;
  private readonly versionSupported;
  constructor(private readonly options: MCPHostOptions) {
    if (!options.serverId || !options.protocolVersion) throw new TypeError('Provide the native server identity and negotiated MCP protocol version');
    this.versionSupported = adapterCompatible('mcp');
    if (!this.versionSupported && options.runtime.mode !== 'off') options.runtime.decline('unsupported_version');
    this.binding = options.runtime.recovery(options.scope);
    this.adapter = { id: 'mcp', version: '0.1.0', framework_version: frameworkVersion('@modelcontextprotocol/sdk') ?? 'unknown', serialization_revision: `mcp-native-${options.protocolVersion}-v1` };
    this.recovery = {
      tool: { name: this.binding.name, description: this.binding.description, inputSchema: structuredClone(this.binding.inputSchema) as Tool['inputSchema'] },
      execute: async (arguments_, options) => {
        if (typeof arguments_.handle !== 'string') throw new TypeError('Recovery requires its scoped handle');
        const page = await this.binding.execute({ ...arguments_, handle: arguments_.handle }, options?.signal ? { signal: options.signal } : undefined);
        return { content: [{ type: 'text', text: JSON.stringify(page) }] };
      },
    };
    this.recoveryExecutor = this.recovery.execute;
    this.recoveryDefinition = JSON.stringify(this.recovery.tool);
  }

  register(tools: readonly MCPToolBinding[]): MCPToolBinding[] {
    return !this.versionSupported || this.options.runtime.mode !== 'compress' || tools.some(item => item.tool.name === this.recovery.tool.name) ? [...tools] : [...tools, this.recovery];
  }

  /** Persist the original result; pass only this copied view to the model.
   * The host supplies its original append-only context manifest across resumes.
   * Final provider serialization and dispatch are outside this result boundary. */
  async projectResult(result: CallToolResult, options: {
    tool: Tool; callId: string; contextManifest: readonly ManifestItem[];
    registeredTools?: readonly MCPToolBinding[]; sequence?: number; signal?: AbortSignal;
  }): Promise<CallToolResult> {
    const report = (reason: string) => this.options.runtime.report(null, { reason, adapter: this.adapter.id });
    if (options.signal?.aborted && !currentOwner()) report('cancelled');
    options.signal?.throwIfAborted();
    if (currentOwner()) return result;
    if (!this.versionSupported || this.options.runtime.mode === 'off') { report('unsupported_version'); return result; }
    if (!plain(result) || !plain(options.tool) || !Array.isArray(result.content) ||
      result.isError || ('resultType' in result && result.resultType !== 'complete') || 'structuredContent' in result || options.tool.outputSchema !== undefined || options.tool.name.startsWith('caveman_')) {
      report('protected_result'); return result;
    }
    const candidates = [], indices = new Map<string, number>();
    for (let i = 0; i < result.content.length; i++) {
      const part = result.content[i];
      if (!plain(part) || part.type !== 'text' || typeof part.text !== 'string' || part.annotations?.audience && !part.annotations.audience.includes('assistant')) continue;
      const key = `mcp-${await sha256(JSON.stringify([this.options.serverId, options.tool.name, options.callId, i]))}`;
      indices.set(key, i);
      candidates.push({ id: key, sourceId: key, content: part.text });
    }
    if (!candidates.length) { report('no_candidate'); return result; }
    const registered = () => {
      const matching = options.registeredTools?.filter(item => item.tool.name === this.binding.name) ?? [];
      return this.options.runtime.ownsBinding(this.binding, this.options.scope) && matching.length === 1 && matching[0] === this.recovery &&
        this.recovery.execute === this.recoveryExecutor && JSON.stringify(this.recovery.tool) === this.recoveryDefinition;
    };
    const bound = registered();
    const outcome = await this.options.runtime.optimize({ scope: this.options.scope, adapter: this.adapter,
      candidates, manifest: options.contextManifest,
      ...(options.sequence !== undefined ? { sequence: options.sequence } : {}), ...(options.signal ? { signal: options.signal } : {}),
      ...(bound ? { binding: this.binding } : {}), recoveryOverheadText: JSON.stringify(this.recovery.tool) });
    if (!outcome.replacements.length) { this.options.runtime.report(outcome); return result; }
    if (bound && !registered()) { report('recovery_unavailable'); return result; }
    if (!outcome.replacements.every(replacement => indices.has(replacement.segment_id))) { report('invalid_plan'); return result; }
    const content = [...result.content];
    for (const replacement of outcome.replacements) {
      const i = indices.get(replacement.segment_id)!;
      content[i] = { ...content[i], text: replacement.text } as typeof content[number];
    }
    const view = { ...result, content };
    this.options.runtime.report(outcome);
    return view;
  }
}
